server/models.py

```diff
@@ -254,43 +247,136 @@ class GroupMember(db.Model, TimestampMixin):
 
-    Members of a group can view each other's submissions.
+    Groups are created when a member not in a group invites another member.
+    Invited members may accept or decline invitations. Active members may
+    revoke invitations and remove members (including themselves).
 
-    Specification:
-    https://github.com/Cal-CS-61A-Staff/ok/wiki/Group-&-Submission-Consistency
+    A group must have at least 1 active member and at lease 2 participants.
+    Degenerates groups are deleted.
     """
     id = db.Column(db.Integer(), primary_key=True)
-    assignment = db.Column(db.ForeignKey("assignment.id"), nullable=False)
-    members = db.relationship("GroupMember")
+    assignment_id = db.Column(db.ForeignKey("assignment.id"), nullable=False)
+
+    assignment = db.relationship("Assignment")
+
+    def size(self):
+        return GroupMember.query.filter_by(group=self).count()
+
+    def has_status(self, user, status):
+        return GroupMember.query.filter_by(
+            user=user,
+            group=self,
+            status=status
+        ).count() > 0
+
+    @staticmethod
+    def lookup(user, assignment):
+        member = GroupMember.query.filter_by(
+            user=user,
+            assignment=assignment
+        ).one_or_none()
+        if member:
+            return member.group
+
+    @staticmethod
+    def invite(sender, recipient, assignment):
+        """Invite a user to a group, creating a group if necessary."""
+        if not assignment.active:
+            raise BadRequest('The assignment is past due')
+        group = Group.lookup(sender, assignment)
+        if not group:
+            group = Group(assignment=assignment)
+            db.session.add(group)
+            group._add_member(sender, 'active')
+        elif not group.has_status(sender, 'active'):
+            raise BadRequest('You are not in the group')
+        group._add_member(recipient, 'pending')
+
+    def remove(self, user, target_user):
+        """Remove a user from the group.
+
+        The user must be an active member in the group, and the target user
+        must be an active or pending member. You may remove yourself to leave
+        the group. The assignment must also be active.
+        """
+        if not self.assignment.active:
+            raise BadRequest('The assignment is past due')
+        if not self.has_status(user, 'active'):
+            raise BadRequest('You are not in the group')
+        self._remove_member(target_user)
+
+    def accept(self, user):
+        """Accept an invitation."""
+        if not self.assignment.active:
+            raise BadRequest('The assignment is past due')
+        member = GroupMember.query.filter_by(
+            user=user,
+            group=self,
+            status='pending'
+        ).one_or_none()
+        if not member:
+            raise BadRequest('{0} is not invited to this group'.format(user.email))
+        member.status = 'active'
```

Comment:
No. It says (in the sqlalchemy docs)[http://docs.sqlalchemy.org/en/latest/orm/session_basics.html#when-do-i-construct-a-session-when-do-i-commit-it-and-when-do-i-close-it] to do the commit somewhere else, so maybe it can go in the controllers. I'm fine with committing in the "public" methods (accept, decline) etc. though